esphome/components/max44009/max44009.cpp

```diff
@@ -0,0 +1,145 @@
+#include "max44009.h"
+
+#include "esphome/core/log.h"
+
+namespace esphome {
+namespace max44009 {
+
+static const char *const TAG = "max44009.sensor";
+
+// REGISTERS
+static const uint8_t MAX44009_REGISTER_CONFIGURATION = 0x02;
+static const uint8_t MAX44009_LUX_READING_HIGH = 0x03;
+static const uint8_t MAX44009_LUX_READING_LOW = 0x04;
+// CONFIGURATION MASKS
+static const uint8_t MAX44009_CFG_CONTINUOUS = 0x80;
+// ERROR CODES
+static const uint8_t MAX44009_OK = 0;
+static const uint8_t MAX44009_ERROR_WIRE_REQUEST = -10;
+static const uint8_t MAX44009_ERROR_OVERFLOW = -20;
+static const uint8_t MAX44009_ERROR_HIGH_BYTE = -30;
+static const uint8_t MAX44009_ERROR_LOW_BYTE = -31;
+
+void MAX44009Sensor::setup() {
+  ESP_LOGCONFIG(TAG, "Setting up MAX44009...");
+  bool state_ok = false;
+  if (this->mode_ == MAX44009Mode::MAX44009_MODE_LOW_POWER) {
+    state_ok = this->set_low_power_mode();
+  } else if (this->mode_ == MAX44009Mode::MAX44009_MODE_CONTINUOUS) {
+    state_ok = this->set_continuous_mode();
+  } else {
+    /*
+     * Mode AUTO: Set mode depending on update interval
+     * - On low power mode, the IC measures lux intensity only once every 800ms
+     * regardless of integration time
+     * - On continuous mode, the IC continuously measures lux intensity
+     */
+    if (this->get_update_interval() < 800) {
+      state_ok = this->set_continuous_mode();
+    } else {
+      state_ok = this->set_low_power_mode();
+    }
+  }
+  if (!state_ok)
+    this->mark_failed();
+}
+
+void MAX44009Sensor::dump_config() {
+  ESP_LOGCONFIG(TAG, "MAX44009:");
+  LOG_I2C_DEVICE(this);
+  if (this->is_failed()) {
+    ESP_LOGE(TAG, "Communication with MAX44009 failed!");
+  }
+}
+
+float MAX44009Sensor::get_setup_priority() const { return setup_priority::DATA; }
+
+void MAX44009Sensor::update() {
+  // update sensor illuminance value
+  float lux = this->read_illuminance_();
+  if (this->error_ != MAX44009_OK) {
+    this->status_set_error();
```

Comment:
```suggestion
    this->status_set_warning();
```

We only use error for things that make a device fully inoperable (like a lost wifi/api/mqtt connection)